// @unibeautify/beautifier-prettydiff/dist/test/index.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const unibeautify_1 = require("unibeautify");
const src_1 = require("../src");
test("should successfully install beautifier", () => {
    const unibeautify = unibeautify_1.newUnibeautify();
    unibeautify.loadBeautifier(src_1.default);
    expect(unibeautify.loadedBeautifiers.map(curr => curr.name)).toEqual([
        src_1.default.name,
    ]);
});
test("should successfully beautify JavaScript text with 2 space indentation", () => {
    const unibeautify = unibeautify_1.newUnibeautify();
    unibeautify.loadBeautifier(src_1.default);
    const text = `function test(n){return n+1;}`;
    const beautifierResult = `function test(n) {\n  return n + 1;\n}`;
    return unibeautify
        .beautify({
        languageName: "JavaScript",
        options: {
            JavaScript: {
                indent_char: " ",
                indent_size: 2,
            },
        },
        text,
    })
        .then(results => {
        expect(results).toBe(beautifierResult);
    });
});
test("should successfully beautify JavaScript text with double quotes", () => {
    const unibeautify = unibeautify_1.newUnibeautify();
    unibeautify.loadBeautifier(src_1.default);
    const text = `console.log('hello world');`;
    const beautifierResult = `console.log("hello world");`;
    return unibeautify
        .beautify({
        languageName: "JavaScript",
        options: {
            JavaScript: {
                indent_char: " ",
                indent_size: 2,
                quotes: "double",
            },
        },
        text,
    })
        .then(results => {
        expect(results).toBe(beautifierResult);
    });
});
