// @unibeautify/beautifier-prettydiff/dist/test/languages/jsx.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const unibeautify_1 = require("unibeautify");
const src_1 = require("../../src");
test("should successfully beautify JSX text", () => {
    const unibeautify = unibeautify_1.newUnibeautify();
    unibeautify.loadBeautifier(src_1.default);
    const text = `export default class TestCase extends React.Component{render(){return\
( <div className={this.props.className} someAttr><div>Smth</div></div>);}}`;
    const beautifierResult = `export default class TestCase extends React.Component {\n\trender() {\n\t\treturn \
(<div className={this.props.className} someAttr="someAttr">\n\t\t\t<div>Smth</div>\n\t\t</div>);\n\t}\n}`;
    return unibeautify
        .beautify({
        languageName: "JSX",
        options: {
            JSX: {
                indent_style: "tab",
                indent_size: 1,
            },
        },
        text,
    })
        .then(results => {
        expect(results).toBe(beautifierResult);
    });
});
